packages/libs/sdk/src/api/api.ts

```diff
@@ -89,65 +37,66 @@ export class API {
 
     this.graphApiKey = graphApiKey;
     this.additionalHeaders = additionalHeaders;
-    this.apolloClient = this.createApolloClient();
-    this.customApolloClient = new CustomApolloClient(this.apolloClient);
+    this.urqlClient = this.createUrqlClient();
+    this.customUrqlClient = new CustomUrqlClient(this.urqlClient);
     this.defaultChain = defaultChain || DEFAULT_CHAIN;
-    this.nfts = new NftsController(this.customApolloClient, this.defaultChain);
+    this.nfts = new NftsController(this.customUrqlClient, this.defaultChain);
     this.tokens = new TokensController(
-      this.customApolloClient,
+      this.customUrqlClient,
       this.defaultChain
     );
-    // Re-export the apolloClient configured to use the Graph API for use with custom queries
-    this.graphApiClient = this.apolloClient;
+    // Re-export the Urql client configured to use the Graph API for use with custom queries
+    this.graphApiClient = this.urqlClient;
   }
 
-  private createApolloClient(): ApolloClient<NormalizedCacheObject> {
-    const authLink = setContext(async (_, { headers }) => {
-      return {
-        headers: {
-          ...headers,
-          ...{ 'x-api-key': this.graphApiKey },
-          ...this.additionalHeaders,
-        },
-      };
-    });
-
-    const cacheStructure = new InMemoryCache({
-      possibleTypes: generatedPossibleTypes.possibleTypes,
-      typePolicies: {
-        EVMSchemaType: {
-          // Always merge EVMSchemaType objects
-          // Added because of warning: "Cache data may be lost when replacing the ethereum field of a Query object."
-          merge: true,
-        },
-        NFT: {
-          keyFields: ['contractAddress', 'tokenId'],
-        },
-        Collection: {
-          keyFields: ['address'],
-        },
-        Contract: {
-          keyFields: ['address'],
-        },
-        TokenEvent: {
-          keyFields: ['transactionHash', 'transferIndex'],
-        },
-        Transaction: {
-          keyFields: ['hash'],
-        },
-        TrendingCollection: {
-          keyFields: ['collection', ['address']],
-        },
-        Wallet: {
-          keyFields: ['address'],
-        },
+  private createUrqlClient(): Client {
+    const headers = { ...this.additionalHeaders };
+    if (this.graphApiKey) headers['x-api-key'] = this.graphApiKey;
+    const useNftKey = (data: Data) =>
+      `${data['contractAddress']}:${data['tokenId']}`;
+    const useAddressAsKey = (data: Data) => `${data['address']}`;
+    const useTransactionHashAndIndex = (data: Data) =>
+      `${data['transactionHash']}:${data['transferIndex']}`;
+    const urqlCache = cacheExchange({
+      schema,
```

Comment:
I added introspection, which is generated by codegen, [see urql docs here](https://formidable.com/open-source/urql/docs/graphcache/schema-awareness/)